tests/test_asyncio.py

```diff
@@ -0,0 +1,178 @@
+        return self.get_profiling_outputs()
+
+    async def profile_tasks_coroutine(self):
+        yappi.start()
+        tasks = self.get_tasks()
+        await asyncio.gather(*tasks)
+        yappi.stop()
+
+    def get_profiling_outputs(self):
+        fstats = yappi.get_func_stats(filter={"name": self.task_to_profile_name})
+        stats = fstats.pop()
+        self.assertEqual(stats.ncall, self.task_count * self.context_switch_count)
+
+        averaged_times = [stats.ttot / self.task_count for _ in range(self.task_count)]
+        return averaged_times
+
+    def get_tasks(self):
+        return [self.task_to_profile(self.duration) for _ in range(self.task_count)]
+
+    @staticmethod
+    async def task_to_profile(duration: float):
+        await asyncio.sleep(duration)
+
+
+class ContextVarsMixin:
+    task_count: int
+    duration: float
+    task_to_profile_name: str
+    context_switch_count: int
+
+    async def profile_tasks_coroutine(self):
+        yappi.set_context_id_callback(get_async_context_id)
+        await super().profile_tasks_coroutine()
+
+    def get_profiling_outputs(self):
+        ttots = []
+        for i in range(self.task_count):
+            fstats = yappi.get_func_stats(filter={"ctx_id": i, "name": self.task_to_profile_name})
+            stats = fstats.pop()
+            self.assertEqual(stats.ncall, self.context_switch_count)
+            ttots.append(stats.ttot)
+        return ttots
+
+    def get_tasks(self):
+        async def await_with_context_id(context_id, awaitable):
+            async_context_id.set(context_id)
+            return await awaitable
+
+        return [await_with_context_id(i, self.task_to_profile(self.duration)) for i in range(self.task_count)]
+
+
+class ThreadTaskMixin:
+    run_in_threadpool_copy_contextvars = False
+    task_to_profile_name = "_task_in_thread"
+    context_switch_count = 1
+
+    async def task_to_profile(self, duration: float):
+        await run_in_threadpool(self.run_in_threadpool_copy_contextvars, self._task_in_thread, duration)
+
+    @staticmethod
+    def _task_in_thread(duration: float):
+        time.sleep(duration)
+
+
+class AsyncUsageWithContextvars(ContextVarsMixin, AsyncUsage):
+    pass
+
+
+class AsyncUsageThreaded(ThreadTaskMixin, AsyncUsage):
+    pass
+
+
+class AsyncUsageThreadedWithContextvars(ThreadTaskMixin, ContextVarsMixin, AsyncUsage):
+    run_in_threadpool_copy_contextvars = True
+
+
+class AsyncUsageThreadedWithContextvarsSelfContained(utils.YappiUnitTestCase):
+    task_count = 5
+    context_switch_count = 1
+    duration = 0.1
+    task_to_profile_name = "_task_in_thread"
+
+    def test_async_cpu(self):
+        ttots = self.profile_tasks("cpu")
+        for ttot in ttots:
+            self.assertLessEqual(ttot, self.duration * 0.1)
+
+    def test_async_wall(self):
+        ttots = self.profile_tasks("wall")
+        for ttot in ttots:
+            self.assertGreaterEqual(ttot, self.duration)
```

Comment:
... and this failure here:

```
Exception: 0.00013099999999999999 not greater than or equal to 0.1
```

I would have at least expected the wall time to be larger than the cpu time, even if something was being tracked wrong.

Not sure what is going on here.

Reply:
I also confirmed that the five values in `ttots` are all basically equal to each other with both clock types.

Reply:
Also, just to be clear, things behave as expected if I drop contextvars and keep threading, or drop threading and keep contextvars.

Note: The contextvar + thread case is critical for profiling starlette / fastapi, as any endpoints that make use of blocking IO calls would be called following the pattern from these tests.